src/runtime/virtcontainers/kata_agent.go

```diff
@@ -766,19 +766,20 @@ func (k *kataAgent) startSandbox(ctx context.Context, sandbox *Sandbox) error {
 
 	if sandbox.config.HypervisorType == RemoteHypervisor {
 		ctx = context.WithValue(ctx, customRequestTimeoutKey, remoteRequestTimeout)
-	} else {
-		// Check grpc server is serving
-		if err = k.check(ctx); err != nil {
+	}
+
+	// If a Policy has been specified, send it to the agent.
+	if len(sandbox.config.AgentConfig.Policy) > 0 {
+		if err := sandbox.agent.setPolicy(ctx, sandbox.config.AgentConfig.Policy); err != nil {
 			return err
 		}
+	}
 
-		// If a Policy has been specified, send it to the agent.
-		if len(sandbox.config.AgentConfig.Policy) > 0 {
-			if err := sandbox.agent.setPolicy(ctx, sandbox.config.AgentConfig.Policy); err != nil {
-				return err
-			}
+	if sandbox.config.HypervisorType != RemoteHypervisor { {
+		// Check grpc server is serving
+		if err = k.check(ctx); err != nil {
```

Comment:
At least at a quick look, it's odd to move the "Check grpc server is serving" code after SetPolicy, because SetPolicy needs the grpc server to be working.

This code suggests to me two possibilities:
1. The k.check() call is not needed. Maybe calling SetPolicy and dealing with a possible error is good enough? If that's the case, perhaps we should remove the k.check() call.
2. The k.check() is needed before SetPolicy, so we should keep the original order of calls.

Would the k.check() call benefit the remote hypervisor code path too, before calling SetPolicy?

Reply:
Good point, let me have a try in a dev environment and see if there are any issues with calling `k.check()` for remote hypervisor. Assuming not then I'm leaning towards running it for both before calling `SetPolicy` for safety.

Reply:
Just to circle back on this - the check code doesn't seem to have any issues in my development environment, so I think it's work keeping.

Reply:
Thanks so much for the valuable input in the review @danmihai1!